From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Mon, 12 Aug 2024 10:37:33 +0800
Subject: [PATCH] Try fixing folia #273


diff --git a/src/main/java/org/bukkit/inventory/CraftingRecipe.java b/src/main/java/org/bukkit/inventory/CraftingRecipe.java
index e8c3afda92d4ae5430d622ea18500985d6cc00f2..c78c1ee19a0960c6eca685a3b0d1ddbcb2831e94 100644
--- a/src/main/java/org/bukkit/inventory/CraftingRecipe.java
+++ b/src/main/java/org/bukkit/inventory/CraftingRecipe.java
@@ -102,7 +102,7 @@ public abstract class CraftingRecipe implements Recipe, Keyed {
     @ApiStatus.Internal
     @NotNull
     protected static ItemStack checkResult(@NotNull ItemStack result) {
-        Preconditions.checkArgument(result.isEmpty(), "Recipe cannot have an empty result"); // Paper
+        Preconditions.checkArgument(!result.isEmpty(), "Recipe cannot have an empty result"); // Paper
         return result;
     }
 }
