From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Tue, 30 Jul 2024 17:58:25 +0800
Subject: [PATCH] FoliaPR Add TPS From Region


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 63eb35aebc4551355ea1ea5cd1305d93502f0dbc..de5922397c7d424715e812819b14030639211846 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2405,6 +2405,28 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Folia start
+    /**
+     * Gets the current location TPS.
+     *
+     * @param location the location for which to get the TPS
+     * @return current location TPS (5s, 15s, 1m, 5m, 15m in Folia-Server), or null if the region doesn't exist
+     */
+    public double @Nullable [] getTPS(Location location) {
+        return server.getTPS(location);
+    }
+
+    /**
+     * Gets the current chunk TPS.
+     *
+     * @param chunk the chunk for which to get the TPS
+     * @return current chunk TPS (5s, 15s, 1m, 5m, 15m in Folia-Server), or null if the region doesn't exist
+     */
+    public double @Nullable [] getTPS(Chunk chunk){
+        return server.getTPS(chunk);
+    }
+    // Folia end
+
     /**
      * Get the advancement specified by this key.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8a7ba7d14f20b8169854daaf26e8f4da4563f166..11025aaab88f109f03cd6cda897fc80488cee060 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2058,6 +2058,24 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     double getAverageTickTime();
     // Paper end
 
+    // Folia start
+    /**
+     * Gets the current location TPS.
+     *
+     * @param location the location for which to get the TPS
+     * @return current location TPS (5s, 15s, 1m, 5m, 15m in Folia-Server), or null if the region doesn't exist
+     */
+    public double @Nullable [] getTPS(Location location);
+
+    /**
+     * Gets the current chunk TPS.
+     *
+     * @param chunk the chunk for which to get the TPS
+     * @return current chunk TPS (5s, 15s, 1m, 5m, 15m in Folia-Server), or null if the region doesn't exist
+     */
+    public double @Nullable [] getTPS(Chunk chunk);
+    // Folia end
+
     // Paper start
     /**
      * Gets the active {@link org.bukkit.command.CommandMap}
