From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <79621885+MrHua269@users.noreply.github.com>
Date: Sat, 23 Dec 2023 21:46:22 +0800
Subject: [PATCH] Fix build


diff --git a/build.gradle.kts b/build.gradle.kts
index 03abea227c555decd559dd58b33913127f3549a4..44b3fc01d8fdce55cffe77b7b02b07b95e182870 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -28,7 +28,7 @@ repositories {
 dependencies {
     extraRuntime(platform("net.kyori:adventure-bom:4.15.0-SNAPSHOT"))
     // Folia start
-    implementation(project(":folia-api"))
+    implementation(project(":luminol-api"))//Luminol
     implementation("io.papermc.paper:paper-mojangapi:${project.version}") {
         exclude("io.papermc.paper", "paper-api")
     }
@@ -88,7 +88,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Folia-$implementationVersion", // Folia
+            "Implementation-Version" to "git-Luminol-$implementationVersion", // Folia //Luminol
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
