From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 31 Jul 2024 14:50:23 +0800
Subject: [PATCH] Threaded region start tick and finished tick event


diff --git a/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java b/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 4bfa83a6683970657eaaaaa265855f0d2e79a3a9..6bbc9a0fb94980cd201d7ddb7cb440fcee7786f4 100644
--- a/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -442,7 +442,7 @@ public final class TickRegionScheduler {
 
             final long scheduledStart = this.getScheduledStart();
             final long scheduledEnd = scheduledStart + TIME_BETWEEN_TICKS;
-
+            new me.earthme.luminol.api.events.TickRegionStartTickEvent(this.region == null ? -1 : this.region.region.id,System.nanoTime()); //Luminol - Threaded regions API
             synchronized (this) {
                 this.currentTickData = new TickTime(
                     lastTickStart, scheduledStart, tickStart, cpuStart,
@@ -477,6 +477,7 @@ public final class TickRegionScheduler {
                 );
 
                 this.addTickTime(time);
+                new me.earthme.luminol.api.events.TickRegionFinishedTickEvent(this.region == null ? -1 : this.region.region.id,System.nanoTime()); //Luminol - Threaded regions API
                 TickRegionScheduler.setTickTask(null);
                 if (this.region != null) {
                     TickRegionScheduler.setTickingRegion(null);
