From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 31 Jul 2024 14:51:16 +0800
Subject: [PATCH] Fix-MC-2025


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index ea1098d2ff613ff634c710b6be6994215719391f..3a0279991d1f65f65aeb8e228f04202761b3e85b 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2623,6 +2623,16 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 nbttagcompound.putBoolean("Paper.FreezeLock", true);
             }
             // Paper end
+
+            // Luminol start - Fix MC-2025
+            AABB boundingBox = getBoundingBox();
+            ListTag boundingBoxList = new ListTag();
+            for (double coord : new double[]{boundingBox.minX, boundingBox.minY, boundingBox.minZ, boundingBox.maxX, boundingBox.maxY, boundingBox.maxZ}) {
+                boundingBoxList.add(DoubleTag.valueOf(coord));
+            }
+            nbttagcompound.put("Luminol.BoundingBox", boundingBoxList);
+            // Luminol end
+
             return nbttagcompound;
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Saving entity NBT");
@@ -2700,6 +2710,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         this.reapplyPosition();
                     }
 
+                    // Luminol start - Fix MC-2025
+                    if (nbt.contains("Luminol.BoundingBox", net.minecraft.nbt.Tag.TAG_LIST)) {
+                        ListTag boundingBoxList = nbt.getList("Luminol.BoundingBox", net.minecraft.nbt.Tag.TAG_DOUBLE);
+                        setBoundingBox(new AABB(boundingBoxList.getDouble(0), boundingBoxList.getDouble(1), boundingBoxList.getDouble(2), boundingBoxList.getDouble(3), boundingBoxList.getDouble(4), boundingBoxList.getDouble(5)));
+                    }
+                    // Luminol end
+
                 } else {
                     throw new IllegalStateException("Entity has invalid rotation");
                 }
